File: catalog.py


### 1. `catalog.py` (Python)

```python
# catalog.py — Python версия

import json
import os
import re
from datetime import datetime
from colorama import init, Fore, Style

init(autoreset=True)
DATA_FILE = "books.json"

class Book:
    def __init__(self, id, isbn, title, author, year, publisher):
        self.id = id
        self.isbn = isbn
        self.title = title
        self.author = author
        self.year = year
        self.publisher = publisher

    def to_dict(self):
        return {"id": self.id, "isbn": self.isbn, "title": self.title,
                "author": self.author, "year": self.year, "publisher": self.publisher}

    @classmethod
    def from_dict(cls, data):
        return cls(data["id"], data["isbn"], data["title"], data["author"],
                   data["year"], data["publisher"])

class BookCatalog:
    def __init__(self):
        self.books = []
        self.load()

    def load(self):
        if os.path.exists(DATA_FILE):
            try:
                with open(DATA_FILE, 'r', encoding='utf-8') as f:
                    data = json.load(f)
                    self.books = [Book.from_dict(b) for b in data]
            except:
                self.books = []

    def save(self):
        with open(DATA_FILE, 'w', encoding='utf-8') as f:
            json.dump([b.to_dict() for b in self.books], f, indent=2, ensure_ascii=False)

    def is_valid_isbn(self, isbn):
        """Проверка ISBN-10 и ISBN-13."""
        isbn = isbn.replace('-', '').replace(' ', '')
        if len(isbn) == 10:
            return self._check_isbn10(isbn)
        elif len(isbn) == 13:
            return self._check_isbn13(isbn)
        return False

    def _check_isbn10(self, isbn):
        if not isbn[:-1].isdigit():
            return False
        total = 0
        for i in range(9):
            total += (i + 1) * int(isbn[i])
        check = isbn[9]
        if check == 'X':
            total += 10 * 10
        elif check.isdigit():
            total += 10 * int(check)
        else:
            return False
        return total % 11 == 0

    def _check_isbn13(self, isbn):
        if not isbn.isdigit():
            return False
        total = 0
        for i in range(13):
            if i % 2 == 0:
                total += int(isbn[i])
            else:
                total += 3 * int(isbn[i])
        return total % 10 == 0

    def add_book(self, isbn, title, author, year, publisher):
        if not self.is_valid_isbn(isbn):
            raise ValueError("Неверный ISBN")
        # Проверка на уникальность ISBN
        for b in self.books:
            if b.isbn == isbn:
                raise ValueError("Книга с таким ISBN уже существует")
        id = len(self.books) + 1
        book = Book(id, isbn, title, author, year, publisher)
        self.books.append(book)
        self.save()
        return id

    def list_all(self):
        if not self.books:
            print(Fore.YELLOW + "Каталог пуст.")
            return
        print(Fore.CYAN + f"{'ID':<4} {'ISBN':<15} {'Название':<30} {'Автор':<20} {'Год':<6} {'Издательство':<20}")
        print("-" * 100)
        for b in self.books:
            print(f"{b.id:<4} {b.isbn:<15} {b.title[:30]:<30} {b.author[:20]:<20} {b.year:<6} {b.publisher[:20]:<20}")

    def search_by_isbn(self, isbn):
        isbn = isbn.replace('-', '').replace(' ', '')
        for b in self.books:
            if b.isbn == isbn:
                return b
        return None

    def search_by_text(self, text):
        text = text.lower()
        results = [b for b in self.books if text in b.title.lower() or text in b.author.lower()]
        return results

    def delete(self, id):
        for i, b in enumerate(self.books):
            if b.id == id:
                del self.books[i]
                self.save()
                return True
        return False

    def edit(self, id, field, value):
        for b in self.books:
            if b.id == id:
                if field == "isbn":
                    if not self.is_valid_isbn(value):
                        raise ValueError("Неверный ISBN")
                    # Проверка уникальности
                    for other in self.books:
                        if other.id != id and other.isbn == value:
                            raise ValueError("ISBN уже используется")
                    b.isbn = value
                elif field == "title":
                    b.title = value
                elif field == "author":
                    b.author = value
                elif field == "year":
                    b.year = int(value)
                elif field == "publisher":
                    b.publisher = value
                else:
                    return False
                self.save()
                return True
        return False

    def stats(self):
        if not self.books:
            print("Нет данных.")
            return
        total = len(self.books)
        authors = {}
        years = {}
        for b in self.books:
            authors[b.author] = authors.get(b.author, 0) + 1
            years[b.year] = years.get(b.year, 0) + 1
        print(Fore.CYAN + "📊 Статистика:")
        print(f"  Всего книг: {total}")
        print("  По авторам:")
        for a, c in sorted(authors.items(), key=lambda x: -x[1]):
            print(f"    {a}: {c}")
        print("  По годам:")
        for y, c in sorted(years.items()):
            print(f"    {y}: {c}")

def main():
    catalog = BookCatalog()
    while True:
        print(Fore.CYAN + "\n📚 Book Catalog (ISBN) (Python)")
        print("1. Добавить книгу")
        print("2. Показать все книги")
        print("3. Поиск по ISBN")
        print("4. Поиск по названию/автору")
        print("5. Удалить книгу")
        print("6. Редактировать книгу")
        print("7. Статистика")
        print("8. Выход")
        choice = input("Выберите действие: ").strip()
        if choice == "1":
            isbn = input("ISBN (10 или 13 цифр): ").strip()
            title = input("Название: ").strip()
            author = input("Автор: ").strip()
            year = int(input("Год: ").strip())
            publisher = input("Издательство: ").strip()
            try:
                id = catalog.add_book(isbn, title, author, year, publisher)
                print(Fore.GREEN + f"✅ Книга добавлена (ID: {id})")
            except Exception as e:
                print(Fore.RED + f"❌ Ошибка: {e}")
        elif choice == "2":
            catalog.list_all()
        elif choice == "3":
            isbn = input("Введите ISBN: ").strip()
            book = catalog.search_by_isbn(isbn)
            if book:
                print(f"ID: {book.id}\nISBN: {book.isbn}\nНазвание: {book.title}\nАвтор: {book.author}\nГод: {book.year}\nИздательство: {book.publisher}")
            else:
                print(Fore.YELLOW + "Книга не найдена.")
        elif choice == "4":
            text = input("Введите название или автора: ").strip()
            results = catalog.search_by_text(text)
            if results:
                for b in results:
                    print(f"{b.id}: {b.title} | {b.author} | {b.year}")
            else:
                print(Fore.YELLOW + "Ничего не найдено.")
        elif choice == "5":
            catalog.list_all()
            id = int(input("Введите ID для удаления: ").strip())
            if catalog.delete(id):
                print(Fore.GREEN + "✅ Книга удалена.")
            else:
                print(Fore.RED + "❌ Книга не найдена.")
        elif choice == "6":
            catalog.list_all()
            id = int(input("Введите ID для редактирования: ").strip())
            field = input("Какое поле редактировать (isbn, title, author, year, publisher): ").strip().lower()
            value = input("Новое значение: ").strip()
            try:
                if catalog.edit(id, field, value):
                    print(Fore.GREEN + "✅ Книга обновлена.")
                else:
                    print(Fore.RED + "❌ Не удалось обновить.")
            except Exception as e:
                print(Fore.RED + f"❌ Ошибка: {e}")
        elif choice == "7":
            catalog.stats()
        elif choice == "8":
            print("До свидания!")
            break
        else:
            print(Fore.RED + "Неверный выбор.")

if __name__ == "__main__":
    main()

File: catalog.php
<?php
// catalog.php — PHP версия

$dataFile = 'books.json';

function loadBooks() {
    global $dataFile;
    if (file_exists($dataFile)) {
        $json = file_get_contents($dataFile);
        return json_decode($json, true) ?: [];
    }
    return [];
}

function saveBooks($books) {
    global $dataFile;
    file_put_contents($dataFile, json_encode($books, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE));
}

$books = loadBooks();

function color($text, $code) {
    return "\033[{$code}m{$text}\033[0m";
}

function isValidISBN($isbn) {
    $isbn = str_replace(['-', ' '], '', $isbn);
    if (strlen($isbn) == 10) return checkISBN10($isbn);
    if (strlen($isbn) == 13) return checkISBN13($isbn);
    return false;
}

function checkISBN10($isbn) {
    if (!preg_match('/^\d{9}[\dX]$/', $isbn)) return false;
    $sum = 0;
    for ($i = 0; $i < 9; $i++) {
        $sum += ($i + 1) * intval($isbn[$i]);
    }
    $check = $isbn[9];
    if ($check == 'X') $sum += 100;
    else $sum += 10 * intval($check);
    return $sum % 11 == 0;
}

function checkISBN13($isbn) {
    if (!preg_match('/^\d{13}$/', $isbn)) return false;
    $sum = 0;
    for ($i = 0; $i < 13; $i++) {
        $digit = intval($isbn[$i]);
        $sum += ($i % 2 == 0) ? $digit : 3 * $digit;
    }
    return $sum % 10 == 0;
}

function listAll($books) {
    if (empty($books)) {
        echo color("Каталог пуст.\n", '33');
        return;
    }
    printf(color("%-4s %-15s %-30s %-20s %-6s %-20s\n", '36'), "ID", "ISBN", "Название", "Автор", "Год", "Издательство");
    echo str_repeat("-", 100) . "\n";
    foreach ($books as $b) {
        $title = strlen($b['title']) > 30 ? substr($b['title'], 0, 30) : $b['title'];
        $author = strlen($b['author']) > 20 ? substr($b['author'], 0, 20) : $b['author'];
        $pub = strlen($b['publisher']) > 20 ? substr($b['publisher'], 0, 20) : $b['publisher'];
        printf("%-4d %-15s %-30s %-20s %-6d %-20s\n", $b['id'], $b['isbn'], $title, $author, $b['year'], $pub);
    }
}

function searchByISBN($books, $isbn) {
    $isbn = str_replace(['-', ' '], '', $isbn);
    foreach ($books as $b) {
        if ($b['isbn'] == $isbn) return $b;
    }
    return null;
}

function searchByText($books, $text) {
    $text = strtolower($text);
    $results = [];
    foreach ($books as $b) {
        if (stripos($b['title'], $text) !== false || stripos($b['author'], $text) !== false) {
            $results[] = $b;
        }
    }
    return $results;
}

function deleteBook(&$books, $id) {
    foreach ($books as $i => $b) {
        if ($b['id'] == $id) {
            array_splice($books, $i, 1);
            saveBooks($books);
            return true;
        }
    }
    return false;
}

function editBook(&$books, $id, $field, $value) {
    foreach ($books as &$b) {
        if ($b['id'] == $id) {
            switch ($field) {
                case 'isbn':
                    if (!isValidISBN($value)) throw new Exception("Неверный ISBN");
                    foreach ($books as $other) {
                        if ($other['id'] != $id && $other['isbn'] == $value) throw new Exception("ISBN уже используется");
                    }
                    $b['isbn'] = $value;
                    break;
                case 'title': $b['title'] = $value; break;
                case 'author': $b['author'] = $value; break;
                case 'year': $b['year'] = (int)$value; break;
                case 'publisher': $b['publisher'] = $value; break;
                default: throw new Exception("Неизвестное поле");
            }
            saveBooks($books);
            return true;
        }
    }
    throw new Exception("Книга не найдена");
}

function stats($books) {
    if (empty($books)) {
        echo "Нет данных.\n";
        return;
    }
    $authors = [];
    $years = [];
    foreach ($books as $b) {
        $authors[$b['author']] = ($authors[$b['author']] ?? 0) + 1;
        $years[$b['year']] = ($years[$b['year']] ?? 0) + 1;
    }
    echo color("📊 Статистика:\n", '36');
    echo "  Всего книг: " . count($books) . "\n";
    echo "  По авторам:\n";
    arsort($authors);
    foreach ($authors as $a => $c) {
        echo "    $a: $c\n";
    }
    echo "  По годам:\n";
    ksort($years);
    foreach ($years as $y => $c) {
        echo "    $y: $c\n";
    }
}

function main() {
    global $books;
    while (true) {
        echo "\n" . color("📚 Book Catalog (ISBN) (PHP)\n", '36');
        echo "1. Добавить книгу\n";
        echo "2. Показать все книги\n";
        echo "3. Поиск по ISBN\n";
        echo "4. Поиск по названию/автору\n";
        echo "5. Удалить книгу\n";
        echo "6. Редактировать книгу\n";
        echo "7. Статистика\n";
        echo "8. Выход\n";
        echo "Выберите действие: ";
        $choice = trim(fgets(STDIN));

        switch ($choice) {
            case '1':
                echo "ISBN (10 или 13 цифр): ";
                $isbn = trim(fgets(STDIN));
                echo "Название: ";
                $title = trim(fgets(STDIN));
                echo "Автор: ";
                $author = trim(fgets(STDIN));
                echo "Год: ";
                $year = (int) trim(fgets(STDIN));
                echo "Издательство: ";
                $publisher = trim(fgets(STDIN));
                if (!isValidISBN($isbn)) {
                    echo color("❌ Неверный ISBN.\n", '31');
                    break;
                }
                foreach ($books as $b) {
                    if ($b['isbn'] == $isbn) {
                        echo color("❌ Книга с таким ISBN уже существует.\n", '31');
                        break 2;
                    }
                }
                $id = count($books) + 1;
                $books[] = ['id' => $id, 'isbn' => $isbn, 'title' => $title, 'author' => $author, 'year' => $year, 'publisher' => $publisher];
                saveBooks($books);
                echo color("✅ Книга добавлена (ID: $id)\n", '32');
                break;
            case '2':
                listAll($books);
                break;
            case '3':
                echo "Введите ISBN: ";
                $isbn = trim(fgets(STDIN));
                $book = searchByISBN($books, $isbn);
                if ($book) {
                    echo "ID: {$book['id']}\nISBN: {$book['isbn']}\nНазвание: {$book['title']}\nАвтор: {$book['author']}\nГод: {$book['year']}\nИздательство: {$book['publisher']}\n";
                } else {
                    echo color("Книга не найдена.\n", '33');
                }
                break;
            case '4':
                echo "Введите название или автора: ";
                $text = trim(fgets(STDIN));
                $results = searchByText($books, $text);
                if (empty($results)) {
                    echo color("Ничего не найдено.\n", '33');
                } else {
                    foreach ($results as $b) {
                        echo "{$b['id']}: {$b['title']} | {$b['author']} | {$b['year']}\n";
                    }
                }
                break;
            case '5':
                listAll($books);
                echo "Введите ID для удаления: ";
                $id = (int) trim(fgets(STDIN));
                if (deleteBook($books, $id)) {
                    echo color("✅ Книга удалена.\n", '32');
                } else {
                    echo color("❌ Книга не найдена.\n", '31');
                }
                break;
            case '6':
                listAll($books);
                echo "Введите ID для редактирования: ";
                $id = (int) trim(fgets(STDIN));
                echo "Какое поле редактировать (isbn, title, author, year, publisher): ";
                $field = trim(fgets(STDIN));
                echo "Новое значение: ";
                $value = trim(fgets(STDIN));
                try {
                    editBook($books, $id, $field, $value);
                    echo color("✅ Книга обновлена.\n", '32');
                } catch (Exception $e) {
                    echo color("❌ Ошибка: " . $e->getMessage() . "\n", '31');
                }
                break;
            case '7':
                stats($books);
                break;
            case '8':
                echo "До свидания!\n";
                exit(0);
            default:
                echo color("Неверный выбор.\n", '31');
        }
    }
}

main();
?>
